File: 프로그래머스/3/81303. 표 편집/표 편집.java
import java.util.*;

class Solution {
    public String solution(int n, int k, String[] cmd) {
        int[] prev = new int[n];
        int[] next = new int[n];
        boolean[] removed = new boolean[n];

        for (int i = 0; i < n; i++) {
            prev[i] = i - 1;
            next[i] = i + 1;
        }
        next[n - 1] = -1;

        // 삭제 스택: {idx, prevIdx, nextIdx}
        ArrayDeque<int[]> stack = new ArrayDeque<>();

        int cur = k;

        for (String s : cmd) {
            char op = s.charAt(0);

            if (op == 'U') {
                int x = Integer.parseInt(s.substring(2));
                for (int i = 0; i < x; i++) cur = prev[cur];

            } else if (op == 'D') {
                int x = Integer.parseInt(s.substring(2));
                for (int i = 0; i < x; i++) cur = next[cur];

            } else if (op == 'C') {
                removed[cur] = true;
                int p = prev[cur];
                int nx = next[cur];
                stack.push(new int[]{cur, p, nx});

                // 연결 끊기
                if (p != -1) next[p] = nx;
                if (nx != -1) prev[nx] = p;

                // 선택 이동
                cur = (nx != -1) ? nx : p;

            } else { // 'Z'
                int[] rec = stack.pop();
                int idx = rec[0];
                int p = rec[1];
                int nx = rec[2];

                removed[idx] = false;

                // 연결 복구
                if (p != -1) next[p] = idx;
                if (nx != -1) prev[nx] = idx;
                prev[idx] = p;
                next[idx] = nx;
            }
        }

        StringBuilder sb = new StringBuilder(n);
        for (int i = 0; i < n; i++) sb.append(removed[i] ? 'X' : 'O');
        return sb.toString();
    }
}

File: 프로그래머스/3/81303. 표 편집/표 편집.kt
import java.util.ArrayDeque

class Solution {
    fun solution(n: Int, k: Int, cmd: Array<String>): String {
        val prev = IntArray(n) { it - 1 }
        val next = IntArray(n) { it + 1 }
        next[n - 1] = -1

        val removed = BooleanArray(n)
        val stack = ArrayDeque<IntArray>() // [idx, prevIdx, nextIdx]

        var cur = k

        for (s in cmd) {
            when (s[0]) {
                'U' -> {
                    val x = s.substring(2).toInt()
                    repeat(x) { cur = prev[cur] }
                }
                'D' -> {
                    val x = s.substring(2).toInt()
                    repeat(x) { cur = next[cur] }
                }
                'C' -> {
                    removed[cur] = true
                    val p = prev[cur]
                    val nx = next[cur]
                    stack.addLast(intArrayOf(cur, p, nx))

                    // unlink
                    if (p != -1) next[p] = nx
                    if (nx != -1) prev[nx] = p

                    // move selection
                    cur = if (nx != -1) nx else p
                }
                'Z' -> {
                    val rec = stack.removeLast()
                    val idx = rec[0]
                    val p = rec[1]
                    val nx = rec[2]
                    removed[idx] = false

                    // relink
                    if (p != -1) next[p] = idx
                    if (nx != -1) prev[nx] = idx
                    prev[idx] = p
                    next[idx] = nx
                }
            }
        }

        val sb = StringBuilder(n)
        for (i in 0 until n) sb.append(if (removed[i]) 'X' else 'O')
        return sb.toString()
    }
}